### PR TITLE
Fix some rewriting rules in rewrite.td

### PR DESCRIPTION
**Pull request by tungld**
Some rules in `rewrite.td`, such as those for Squeeze, Unsqueeze, and Reshape, require tensors having shape and rank. This patch adds conditions for such rules.

With this patch, we are able to compile yolov3 in the onnx model zoo.

### CODE REVIEW

File: src/Dialect/ONNX/ONNXOpsHelper.hpp [hunk 1]
```diff
@@ -142,6 +142,9 @@ bool HasSpecifiedConstantShape(mlir::Value value, mlir::Value shape);
 bool AreTheSameConstantOpDenseAttr(
     mlir::Builder &builder, int64_t rank, mlir::Value lhsOp, mlir::Value rhsOp);
 
+/// Test if 'val' has shape and rank or not.
```
Comment (AlexandreEichenberger) [inline src/Dialect/ONNX/ONNXOpsHelper.hpp:145]: Thanks for moving that op to the helper header file.

**Review by AlexandreEichenberger — APPROVED** [review]
The code LGTM.

I will defer to @chentong319 as I know he has a bigger picture view of these frontend optimizations.

**Comment by chentong319** [comment]
A possible small trick: when the rank or shape is unknown but only used to handle negative axis value, we can assume it to be 0. As a result, we can correctly compare when the two axis values are both positive or negative. We just conservatively return not the same when one is positive and one is negative.

**Comment by tungld** [comment]
> A possible small trick: when the rank or shape is unknown but only used to handle negative axis value, we can assume it to be 0. As a result, we can correctly compare when the two axis values are both positive or negative. We just conservatively return not the same when one is positive and one is negative.

@chentong319 done. Thanks for the suggestion.

**Review by chentong319 — APPROVED** [review]
LGTM!

**Pull request merged**